readme.txt
=== IP Location Block ===
Contributors: darkog
Donate link:
Tags: country, block, ip address, ip geo block, geolocation, ip
Requires at least: 3.7
Tested up to: 5.7
Stable tag: 1.0.3
License: GPLv3
License URI: https://www.gnu.org/licenses/gpl-3.0.txt

Easily setup location block based on the visitor country by using ip and asn details. Protects your site from spam, login attempts, zero-day exploits, malicious access & more.

== Description ==

IP Location Block plugin that allows you to block access to your site based on the visitor location while also keeping your site safe from malicious attacks. The plugin brings a smart and powerful protection methods named as "**WP Zero-day Exploit Prevention**" and "**WP Metadata Exploit Protection**".

Combined with those methods and IP address geolocation, you'll be surprised to find a bunch of malicious or undesirable access blocked in the logs of this plugin after several days of installation.

**Note:** This plugin is based on the now abandoned "IP Geo Block" plugin by tokkonopapa. I fixed various issues and improved the overall codebase.

= Features =

* **Privacy by design:**  
  IP address is always encrypted on recording in logs/cache. Moreover, it can be anonymized and restricted on sending to the 3rd parties such as geolocation APIs or whois service.

* **Immigration control:**  
  Access to the basic and important entrances into back-end such as `wp-comments-post.php`, `xmlrpc.php`, `wp-login.php`, `wp-signup.php`, `wp-admin/admin.php`, `wp-admin/admin-ajax.php`, `wp-admin/admin-post.php` will be validated by means of a country code based on IP address. It allows you to configure either whitelist or blacklist to [specify the countires](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements "ISO 3166-1 alpha-2 - Wikipedia"), [CIDR notation](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing "Classless Inter-Domain Routing - Wikipedia") for a range of IP addresses and [AS number](https://en.wikipedia.org/wiki/Autonomous_system_(Internet) "Autonomous system (Internet) - Wikipedia") for a group of IP networks.

* **Zero-day Exploit Prevention:**  
  Unlike other security firewalls based on attack patterns (vectors), the original feature "**W**ord**P**ress **Z**ero-day **E**xploit **P**revention" (WP-ZEP) is focused on patterns of vulnerability. It is simple but still smart and strong enough to block any malicious accesses to `wp-admin/*.php`, `plugins/*.php` and `themes/*.php` even from the permitted countries. It will protect your site against certain types of attack such as CSRF, LFI, SQLi, XSS and so on, **even if you have some vulnerable plugins and themes in your site**.

* **Guard against login attempts:**  
  In order to prevent hacking through the login form and XML-RPC by brute-force and the reverse-brute-force attacks, the number of login attempts will be limited per IP address even from the permitted countries.

* **Minimize server load against brute-force attacks:**  
  You can configure this plugin as a [Must Use Plugins](https://codex.wordpress.org/Must_Use_Plugins "Must Use Plugins &laquo; WordPress Codex") so that this plugin can be loaded prior to regular plugins. It can massively [reduce the load on server](https://iplocationblock.com/codex/validation-timing/ "Validation timing | IP Location Block").

* **Prevent malicious down/uploading:**  
  A malicious request such as exposing `wp-config.php` or uploading malwares via vulnerable plugins/themes can be blocked.

* **Block badly-behaved bots and crawlers:**  
  A simple logic may help to reduce the number of rogue bots and crawlers scraping your site.

* **Support of BuddyPress and bbPress:**  
  You can configure this plugin so that a registered user can login as a membership from anywhere, while a request such as a new user registration, lost password, creating a new topic and subscribing comment can be blocked by country. It is suitable for [BuddyPress](https://wordpress.org/plugins/buddypress/ "BuddyPress &mdash; WordPress Plugins") and [bbPress](https://wordpress.org/plugins/bbpress/ "WordPress &rsaquo; bbPress &laquo; WordPress Plugins") to help reducing spams.

* **Referrer suppressor for external links:**  
  When you click an external hyperlink on admin screens, http referrer will be eliminated to hide a footprint of your site.

* **Multiple source of IP Geolocation databases:**  
  [MaxMind GeoLite2 free databases](https://www.maxmind.com "MaxMind - IP Geolocation and Online Fraud Prevention") (it requires PHP 5.4.0+) and [IP2Location LITE databases](https://www.ip2location.com/ "IP Address Geolocation to Identify Website Visitor's Geographical Location") can be installed in this plugin. Also free Geolocation REST APIs and whois information can be available for audit purposes.  
  Father more, [dedicated API class libraries](https://iplocationblock.com/cloudflare-cloudfront-api-class-library/ "CloudFlare & CloudFront API class library | IP Location Block") can be installed for CloudFlare and CloudFront as a reverse proxy service.

* **Customizing response:**  
  HTTP response code can be selectable as `403 Forbidden` to deny access pages, `404 Not Found` to hide pages or even `200 OK` to redirect to the top page.
  You can also have a human friendly page (like `404.php`) in your parent/child theme template directory to fit your site design.

* **Validation logs:**  
  Validation logs for useful information to audit attack patterns can be manageable.

* **Cooperation with full spec security plugin:**  
  This plugin is lite enough to be able to cooperate with other full spec security plugin such as [Wordfence Security](https://wordpress.org/plugins/wordfence/ "Wordfence Security &mdash; WordPress Plugins"). See [this report](https://iplocationblock.com/codex/page-speed-performance/ "Page speed performance | IP Location Block") about page speed performance.

* **Extendability:**  
  You can customize the behavior of this plugin via `add_filter()` with [pre-defined filter hook](https://iplocationblock.com/codex/ "Codex | IP Location Block"). See various use cases in [samples.php](https://iplocationblock.com/codex/example-use-cases-for-the-developer-hooks/) bundled within this package.
  You can also get the extension [IP Geo Allow](https://github.com/ddur/WordPress-IP-Geo-Allow "GitHub - ddur/WordPress-IP-Geo-Allow: WordPress Plugin Exension for WordPress-IP-Geo-Block Plugin") by [Dragan](https://github.com/ddur "ddur (Dragan) - GitHub"). It makes admin screens strictly private with more flexible way than specifying IP addresses.

* **Self blocking prevention and easy rescue:**  
  Website owners do not prefer themselves to be blocked. This plugin prevents such a sad thing unless you force it. And futhermore, if such a situation occurs, you can [rescue yourself](https://iplocationblock.com/codex/what-should-i-do-when-im-locked-out/ "What should I do when I'm locked out? | IP Location Block") easily.

* **Clean uninstallation:**  
  Nothing is left in your precious mySQL database after uninstallation. So you can feel free to install and activate to make a trial of this plugin's functionality.

= Attribution =

This package includes GeoLite2 library distributed by MaxMind, available from [MaxMind](https://www.maxmind.com "MaxMind - IP Geolocation and Online Fraud Prevention") (it requires PHP 5.4.0+), and also includes IP2Location open source libraries available from [IP2Location](https://www.ip2location.com "IP Address Geolocation to Identify Website Visitor's Geographical Location").

Also thanks for providing the following great services and REST APIs for free.

* [http://ip-api.com/](http://ip-api.com/ "IP-API.com - Free Geolocation API") (IPv4, IPv6 / free for non-commercial use)
* [http://geoiplookup.net/](http://geoiplookup.net/ "What Is My IP Address | GeoIP Lookup") (IPv4, IPv6 / free)
* [https://ipinfo.io/](https://ipinfo.io/ "IP Address API and Data Solutions") (IPv4, IPv6 / free)
* [https://ipapi.com/](https://ipapi.com/ "ipapi - IP Address Lookup and Geolocation API") (IPv4, IPv6 / free, need API key)
* [https://ipdata.co/](https://ipdata.co/ "ipdata.co - IP Geolocation and Threat Data API") (IPv4, IPv6 / free, need API key)
* [https://ipstack.com/](https://ipstack.com/ "ipstack - Free IP Geolocation API") (IPv4, IPv6 / free for registered user, need API key)
* [https://ipinfodb.com/](https://ipinfodb.com/ "Free IP Geolocation Tools and API| IPInfoDB") (IPv4, IPv6 / free for registered user, need API key)

= Development =

Development of this plugin is promoted at [WordPress-IP-Geo-Block](https://github.com/tokkonopapa/WordPress-IP-Geo-Block "tokkonopapa/WordPress-IP-Geo-Block - GitHub") and class libraries to handle geo-location database are developed separately as "add-in"s at [WordPress-IP-Geo-API](https://github.com/tokkonopapa/WordPress-IP-Geo-API "tokkonopapa/WordPress-IP-Geo-API - GitHub").

All contributions will always be welcome. Or visit my [development blog](https://iplocationblock.com/ "IP Location Block").

== Installation ==

= Using The WordPress Dashboard =

1. Navigate to the 'Add New' in the plugins dashboard
2. Search for 'IP Location Block'
3. Click 'Install Now'
4. Activate the plugin on the Plugin dashboard
5. Stay cool for a while and go to 'Settings' &raquo; 'IP Location Block'
6. Try 'Best for Back-end' button for easy setup at the bottom of this plugin's setting page.

Please refer to [the document](https://iplocationblock.com/codex/ "Codex | IP Location Block") for your best setup.

== Frequently Asked Questions ==

= Does the site using this plugin comply with GDPR? =

This plugin is designed based on the principle of "Privacy by design" so that you can compliantly run it to GDPR. As guarding against personal data breach, IP addresses in this plugin are encrypted and also can be anonymized by default. It also provides some functions not only to manually erase them but also to automatically remove them when those are exceeded a certain amount/time.

However, these are the part of GDPR requirements and do not guarantee that the site is compliant with GDPR. Refer to [3.0.11 release note](https://iplocationblock.com/changelog/0-3-0-11-release-note/) for details.

= Is there a way to migrate from IP Geo Block"

Yes, if "IP Geo Block" settings are detected, you will see migrate option in the Settings last in "Plugin Settings" section. This will copy the settings from "IP Geo Block" only.

= Does this plugin support multisite? =

Yes. You can synchronize the settings with all the sites on the network when you activate on network and enable "**Network wide settings**" in "**Plugin settings**" section.

= Does this plugin works well with caching? =

The short answer is **YES**, especially for the purpose of security e.g. blocking malicious access both on the back-end and on the front-end.

You can find the long answer and the compatibility list of cache plugins at "[Compatibility with cache plugins](https://iplocationblock.com/codex/compatibility-with-cache-plugins/ 'Compatibility with cache plugins | IP Location Block')".

= I still have access from blacklisted country. Does it work correctly? =

Absolutely, YES.

Sometimes, a WordFence Security user would report this type of claim when he/she found some accesses in its Live traffic view. But please don't worry. Before WordPress runs, WordFence cleverly filters out malicious requests to your site using <a href="https://php.net/manual/en/ini.core.php#ini.auto-prepend-file" title="PHP: Description of core php.ini directives - Manual">auto_prepend_file</a> directive to include PHP based Web Application Firewall. Then this plugin validates the rest of the requests that pass over Wordfence because those were not in WAF rules, especially you enables "**Prevent Zero-day Exploit**".

It would also possibly be caused by the accuracy of country code in the geolocation databases. Actually, there is a case that a same IP address has different country code.

For more detail, please refer to "[I still have access from blacklisted country.](https://iplocationblock.com/codex/i-still-have-access-from-blacklisted-country/ 'I still have access from blacklisted country. | IP Location Block')".

= How can I test this plugin works? =

The easiest way is to use [free proxy browser addon](https://www.google.com/search?q=free+proxy+browser+addon "free proxy browser addon - Google Search").

Another one is to use [http header browser addon](https://www.google.com/search?q=browser+add+on+modify+http+header "browser add on modify http header - Google Search").

You can add an IP address to the `X-Forwarded-For` header to emulate the access behind the proxy. In this case, you should add `HTTP_X_FORWARDED_FOR` into the "**$_SERVER keys for extra IPs**" on "**Settings**" tab.

See more details at "[How to test prevention of attacks](https://iplocationblock.com/?codex-category=test-prevention-of-attacks 'Codex | IP Location Block')".

= I'm locked out! What shall I do? =

Please find the solution in [Quick recovery from blocking on your login page](https://iplocationblock.com/codex/quick-recovery-from-blocking-on-login-page/ "Quick recovery from blocking on your login page | IP Location Block") at first.

You can also find another solution by editing "**Emergent Functionality**" code section near the bottom of `ip-location-block.php`. This code block can be activated by replacing `/*` (opening multi-line comment) at the top of the line to `//` (single line comment), or `*` at the end of the line to `*/` (closing multi-line comment).

`/**
 * Invalidate blocking behavior in case yourself is locked out.
 *
 * How to use: Activate the following code and upload this file via FTP.
 */
/* -- ADD '/' TO THE TOP OR END OF THIS LINE TO ACTIVATE THE FOLLOWINGS -- */
function ip_location_block_emergency( $validate, $settings ) {
    $validate['result'] = 'passed';
    return $validate;
}
add_filter( 'ip-location-block-login', 'ip_location_block_emergency', 1, 2 );
add_filter( 'ip-location-block-admin', 'ip_location_block_emergency', 1, 2 );
// */`

Please not that you have to use an [appropriate editor](https://codex.wordpress.org/Editing_Files#Using_Text_Editors "Editing Files &laquo; WordPress Codex").

After saving and uploading it to `/wp-content/plugins/ip-location-block/` on your server via FTP, you become to be able to login again as an admin.

Remember that you should upload the original one after re-configuration to deactivate this feature.

[This document](https://iplocationblock.com/codex/what-should-i-do-when-im-locked-out/ "What should I do when I'm locked out? | IP Location Block") can also help you.

= Do I have to turn on all the selection to enhance security? =

Yes. Roughly speaking, the strategy of this plugin has been constructed as follows:

- **Block by country**  
  It blocks malicious requests from outside your country.

- **Prevent Zero-day Exploit**  
  It blocks malicious requests from your country.

- **Force to load WP core**  
  It blocks the request which has not been covered in the above two.

- **Bad signatures in query**  
  It blocks the request which has not been covered in the above three.

Please try "**Best for Back-end**" button at the bottom of this plugin's setting page for easy setup. And also see more details in "[The best practice of target settings](https://iplocationblock.com/codex/the-best-practice-for-target-settings/ 'The best practice of target settings | IP Location Block')".

= Does this plugin validate all the requests? =

Unfortunately, no. This plugin can't handle the requests that are not parsed by WordPress. In other words, a standalone file (PHP, CGI or something executable) that is unrelated to WordPress can't be validated by this plugin even if it is in the WordPress install directory.

But there's exceptions: When you enable "**Force to load WP core**" for **Plugins area** or **Themes area**, a standalone PHP file becomes to be able to be blocked. Sometimes this kind of file has some vulnerabilities. This function protects your site against such a case.

= How to resolve "Sorry, your request cannot be accepted."? =

If you encounter this message, please refer to [this document](https://iplocationblock.com/codex/why-sorry-your-request-cannot-be-accepted/ "Why &ldquo;Sorry, your request cannot be accepted&rdquo; ? | IP Location Block") to resolve your blocking issue.

If you can't solve your issue, please let me know about it on the [support forum](https://wordpress.org/support/plugin/ip-location-block/ "View: Plugin Support &laquo;  WordPress.org Forums"). Your logs in this plugin and "**Installation information**" at "**Plugin settings**" will be a great help to resolve the issue.

= How can I fix "Unable to write" error? =

When you enable "**Force to load WP core**" options, this plugin will try to configure `.htaccess` in your `/wp-content/plugins/` and `/wp-content/themes/` directory in order to protect your site against the malicious attacks to the [OMG plugins and themes](https://iplocationblock.com/prevent-exposure-of-wp-config-php/ "Prevent exposure of wp-config.php | IP Location Block").

But some servers doesn't give read / write permission against `.htaccess` to WordPress. In this case, you can configure `.htaccess` files by your own hand instead of enabling "**Force to load WP core**" options.


changelog.txt
= 0.3.0.17.4 =
* **Fix issue:** Add an option "Metadata Exploit Protection" in "Validation rules and behavior" section and make it disabled by default because it potentially caused "Sorry, your request cannot be accepted".
* **Fix issue:** Fix the issue that direct request to PHP files under plugins/themes directory was "404 Not found" when WP-ZEP enabled.
* **Fix issue:** Fix the incorrect link to the corresponding option page on error message in multisite environment.
* **Improvement:** Update the url to the google map api to prevent 301 redirection.
* **Improvement:** Use [RIPE stat Data API](https://stat.ripe.net/docs/data_api "RIPEstat &#8212; Internet Measurements and Analysis") to get whois information instead of using YQL (Yahoo! Query Language) which was retired on Jan. 3, 2019.

= 0.3.0.17.3 =
Sorry for the frequent update.
* **Fix issue:** Remove `wp_user_roles` from "Metadata Exploit Protection" which was equipped in 3.0.17 because some plugins/themes would add their own role and capability without admin authority.

= 0.3.0.17.2 =
* **Fix issue:** Fixed the fatal error under certain condition.

= 0.3.0.17.1 =
* **Fix issue:** Inadequate verification of IPv4 with some CIDR notation.
* **Fix issue:** Insufficient path parsing for document root and content directory when real path and symbolic link were mixtured on virtual host.
* **Fix issue:** Unexpected blocking on admin dashboard when home url is different from site url.
* **Fix issue:** `wp-signup.php` was not blocked when only "Register" was enabled as "Target actions" at "Login form" in "Back-end target settings" section.
* **Improvement:** Improve compatibility with Advanced Access Manager.
* **Improvement:** Use comment syntax `@codingStandardsIgnoreLine` for deprecated function on PHP7+ to avoid false positive by some scanning tool like PHP Compatibility Checker.
Hava a happy new year!

= 0.3.0.17 =
* **Fix issue:** Avoid the error of "Can not load Geolocation API libraries" caused by the race condition on upgrading this plugin.
* **New feature:** Enhance protecting against certain types of attacks by privilege escalation.
* **New feature:** Add a new option of "**Network wide settings**" in "**Plugin settings**" section to synchronize all the settings with the main blog on multisite.
* **New feature:** Add "**Preset filters**" using [`ip-location-block-logs[-preset]`](https://iplocationblock.com/codex/ip-location-block-logs-preset/ "ip-location-block-logs[-preset] | IP Location Block") filter hook to make analyze logs easy on "**Logs**" tab.
* **Improvement:** When `pwd` is set at "**$_POST key to record with value**" in "**Privacy and record settings**" section, mask the password in logs unless the login attempt is blocked.
* **Improvement:** Flush out the buffer of "**Live update**" on "**Logs**" tab when stop button is activated to avoid data remains.
* **Improvement:** Update the geolocation API service "geoip.nekudo.com" to "ipapi.com".
* **Improvement:** Update the class for Maxmind GeoLite Legacy databases to prepare against the upcoming expiry of support.
* See [0.3.0.17 release note](https://iplocationblock.com/changelog/0-3-0-17-release-note/) for some details.

= 0.3.0.16 =
* **Fix issue:** Fix the issue that some functions did not work properly when "Validation timing" was set as `"mu-plugins" (ip-location-block-mu.php)` under some conditions before WP 4.0.
* **New feature:** Add "**Sites list**" page at Network admin page to look through the status of the sites if this plugin is network activated on multisite.
* **Improvement:** Support [Swift Performance Lite](https://wordpress.org/plugins/swift-performance-lite/ "Swift Performance Lite &#124; WordPress.org") cache plugin with "Disk Cashe with PHP" caching mode.
* **Improvement:** Apply simulation mode to not only the front-end but also to the back-end.
* **Improvement:** Optimize SQL issuing to reduce some queries.
* See [0.3.0.16 release note](https://iplocationblock.com/changelog/0-3-0-16-release-note/) for some details.

= 0.3.0.15 =
* **Fix issue:** Fix the issue that "Can not find Geolocation API libraries" error would always appear when once GeoLite2, Maxmind and IP2Location were unselected.
* **Fix issue:** Fix the issue that https in ipstack.com is available only for premiums users.
* **New feature:** Introduce a new filter hook [`ip-location-block-logs-preset`](https://iplocationblock.com/codex/ip-location-block-logs-preset/ "ip-location-block-logs[-preset] | IP Location Block") that adds an UI of "Preset filters" at "Search in logs".
* **Improvement:** Prevent to make a cached page by WP Fastest Cache on blocking when "Validation timing" is "init" action hook.
* **Improvement:** Make the priority of "UA string and qualification" is higher than that of "Block badly-behaved bots and crawlers" on public facing pages.
* **Improvement:** Update google charts based on the [release note](https://developers.google.com/chart/interactive/docs/release_notes "Google Charts Release Notes &nbsp;|&nbsp; Charts | Google Developers").
* **Improvement:** Always show "Google Maps API key" text box in "Plugin settings" section. If it is saved as empty, Google Maps API will no longer be used and a simple map will appear on "Search" tab.
* **Improvement:** Follow the change of RIPE whois format related to AS number.
* **Improvement:** Add UI to select the number of entries per page on IP address tables.
* **Improvement:** Multiple entries on IP address tables can be selected across the pages for bulk action.
* **Improvement:** Multiple Geolocation APIs can be selected on "Search" tab.
* **Improvement:** The country code of admin should not be cached if it's `ZZ` during initialization process.

= 0.3.0.14 =
* **Fix issue:** Fix wrong handling of anonymized IPv6 on clicking a link or applying bulk actions.

samples.php
<?php
/**
 * Samples/Snippets to extend functionality of IP Location Block
 *
 * @package   IP_Location_Block
 * @author    tokkonopapa <tokkonopapa@yahoo.com>
 * @license   GPL-3.0
 * @link      https://iplocationblock.com/
 * @copyright 2013-2019 tokkonopapa
 */
/* Start loading wp-load.php */
if ( ! empty( $_GET['wp-load'] ) ) {
	$res = include_once '../../../wp-load.php';
}
die( isset( $res ) ? '1' : '0' );
/* End of loading wp-load.php */

/**
 * Example 1: Usage of 'ip-location-block-ip-addr'
 * Use case: Replace ip address for test purpose
 *
 * @param  string $ip original ip address
 * @return string $ip replaced ip address
 */
function my_replace_ip( $ip ) {
	return '98.139.183.24'; // yahoo.com
}

add_filter( 'ip-location-block-ip-addr', 'my_replace_ip' );


/**
 * Example 2: Usage of 'ip-location-block-ip-addr'
 * Use case: Retrieve ip address behind the proxy
 *
 * @param  string $ip original ip address
 * @return string $ip replaced ip address
 */
function my_retrieve_ip( $ip ) {
	if ( ! empty( $_SERVER['HTTP_X_FORWARDED_FOR'] ) ) {
		$tmp = explode( ',', $_SERVER['HTTP_X_FORWARDED_FOR'] );
		$tmp = trim( $tmp[0] );
		if ( filter_var( $tmp, FILTER_VALIDATE_IP ) ) {
			$ip = $tmp;
		}
	}

	return $ip;
}

add_filter( 'ip-location-block-ip-addr', 'my_retrieve_ip' );


/**
 * Example 3: Validate ip address before authrization in admin area
 * Use case: When an emergency of yourself being locked out
 *
 */
function my_emergency( $validate ) {
	// password is required even in this case
	$validate['result'] = 'passed';
	return $validate;
}

add_filter( 'ip-location-block-login', 'my_emergency' );
add_filter( 'ip-location-block-admin', 'my_emergency' );


/**
 * Example 4: Usage of 'ip-location-block-comment'
 * Use case: Block comment from specific IP addresses in the blacklist
 *
 * @param  string $validate['ip'] ip address
 * @param  string $validate['code'] country code
 * @return array $validate add 'result' as 'passed' or 'blocked' if possible
 */
function my_blacklist( $validate ) {
	$blacklist = array(
		'123.456.789.',
	);

	foreach ( $blacklist as $ip ) {
		if ( strpos( $ip, $validate['ip'] ) === 0 ) {
			$validate['result'] = 'blocked';
			break;
		}
	}

	return $validate;
}

add_filter( 'ip-location-block-comment', 'my_blacklist' );


/**
 * Example 5: Usage of 'ip-location-block-login' and 'ip-location-block-xmlrpc'
 * Use case: Allow from specific countries in the whitelist
 *
 * @param  string $validate['ip'] ip address
 * @param  string $validate['code'] country code
 * @return array $validate add 'result' as 'passed' or 'blocked' if possible
 */
function my_whitelist( $validate ) {
	$whitelist = array(
		'JP', // should be upper case
	);

	$validate['result'] = 'blocked';

	if ( in_array( $validate['code'], $whitelist, true ) ) {
		$validate['result'] = 'passed';
	}

	return $validate;
}

add_filter( 'ip-location-block-login', 'my_whitelist' );
add_filter( 'ip-location-block-xmlrpc', 'my_whitelist' );


/**
 * Example 6: Validate specific actions of admin-ajax.php at front-end
 * Use case: Give permission to ajax with specific action at public facing page
 *
 * @global array $_GET and $_POST requested queries
 * @param  array $validate validation results
 * @return array $validate add 'result' as 'passed' when 'action' is OK
 */
function my_permitted_ajax( $validate ) {
	$whitelist = array(
		'permitted_action',
	);

	if ( isset( $_REQUEST['action'] ) && in_array( $_REQUEST['action'], $whitelist, true ) )
		$validate['result'] = 'passed';

	return $validate; // should not set 'passed' to validate by country code
}

if ( defined( 'DOING_AJAX' ) && DOING_AJAX )
	add_filter( 'ip-location-block-admin', 'my_permitted_ajax' );


/**
 * Example 7: Validate extra IP addresses with CIDR prior to other validations
 * Use case: Get IPs with CIDR from Amazon AWS and set them to the black list
 *
 * @param  array $extra_ips array of white list and black list
 * @param  string $hook 'comment', 'xmlrpc', 'login', 'admin'
 * @return array $extra_ips updated array
 */
define( 'MY_EXTRA_IPS_LIST', 'my_extra_ips_list' );
define( 'MY_EXTRA_IPS_CRON', 'my_extra_ips_cron' );

function my_extra_ips_get() {
	$list = json_decode(
		@file_get_contents( 'https://ip-ranges.amazonaws.com/ip-ranges.json' ),
		TRUE // convert object to array
	);

	//  keep the list in the cache
	if ( is_array( $list['prefixes'] ) ) {
		$list = implode( ',', array_column( $list['prefixes'], 'ip_prefix' ) );
		set_transient( MY_EXTRA_IPS_LIST, $list, DAY_IN_SECONDS );
	}

	if ( ! wp_next_scheduled( MY_EXTRA_IPS_CRON ) )
		wp_schedule_single_event( time() + HOUR_IN_SECONDS, MY_EXTRA_IPS_CRON );

	return $list;
}

function my_extra_ips_hook( $extra_ips, $hook ) {
	// if the list does not exist, then update
	$list = get_transient( MY_EXTRA_IPS_LIST );

	if ( ! $list )
		wp_schedule_single_event( time(), MY_EXTRA_IPS_CRON );

	// restrict the target hook
	if ( in_array( $hook, array( 'xmlrpc', 'login' ), true ) ) {
		$extra_ips['black_list'] .= ( $extra_ips['black_list'] ? ',' : '' ) . $list;
	}

	return $extra_ips;
}

add_action( MY_EXTRA_IPS_CRON, 'my_extra_ips_get' );
add_filter( 'ip-location-block-extra-ips', 'my_extra_ips_hook', 10, 2 );


/**
 * Example 8: Usage of `ip-location-block-xxxxxx-(status|reason)`
 * Use case: Customize the http response status code and message at blocking
 *
 * @param  int $code or string $msg
 * @return int $code or string $msg
 */
function my_xmlrpc_status( $code ) { return 403; }
function my_login_status ( $code ) { return 503; }
function my_login_reason ( $msg  ) { return "Sorry, this service is unavailable."; }

add_filter( 'ip-location-block-xmlrpc-status', 'my_xmlrpc_status' );
add_filter( 'ip-location-block-login-status',  'my_login_status'  );
add_filter( 'ip-location-block-login-reason',  'my_login_reason'  );


/**
 * Example 9: Usage of 'ip-location-block-bypass-admins'
 * Use case: Specify the admin request with a specific queries to bypass WP-ZEP
 *
 * @param  array $queries array of admin queries which should bypass WP-ZEP.
 * @return array $queries array of admin queries which should bypass WP-ZEP.
 */
function my_bypass_admins( $queries ) {
	// <form method="POST" action="wp-admin/admin-post.php">
	// <input type="hidden" name="action" value="do-my-action" />
	// <input type="hidden" name="page" value="my-plugin-page" />
	// </form>
	$whitelist = array(
		'do-my-action',
		'my-plugin-page',
	);
	return array_merge( $queries, $whitelist );
}

add_filter( 'ip-location-block-bypass-admins', 'my_bypass_admins' );


/**
 * Example 10: Usage of 'ip-location-block-bypass-plugins'
 * Use case: Specify the plugin which should bypass WP-ZEP
 *
 * @param  array of plugin name which should bypass WP-ZEP.
 * @return array of plugin name which should bypass WP-ZEP.
 */
function my_bypass_plugins( $plugins ) {
	// ex) wp-content/plugins/my-plugin/something.php
	$whitelist = array(
		'my-plugin',
	);
	return array_merge( $plugins, $whitelist );
}

add_filter( 'ip-location-block-bypass-plugins', 'my_bypass_plugins' );


/**
 * Example 11: Usage of 'ip-location-block-bypass-themes'
 * Use case: Specify the theme which should bypass WP-ZEP
 *
 * @param  array of theme name which should bypass WP-ZEP.
 * @return array of theme name which should bypass WP-ZEP.
 */
function my_bypass_themes( $themes ) {
	// ex) wp-content/themes/my-theme/something.php
	$whitelist = array(
		'my-theme',
	);
	return array_merge( $themes, $whitelist );
}

add_filter( 'ip-location-block-bypass-themes', 'my_bypass_themes' );


/**
 * Example 12: Usage of 'ip-location-block-headers'
 * Use case: Change the user agent strings when accessing geolocation API
 *
 * Notice: Be careful about HTTP header injection.
 * @param  string $args http request headers for `wp_remote_get()`
 * @return string $args http request headers for `wp_remote_get()`
 */
function my_user_agent( $args ) {
    $args['user-agent'] = 'my user agent strings';
    return $args;
}

add_filter( 'ip-location-block-headers', 'my_user_agent' );


/**
 * Example 13: Usage of 'ip-location-block-maxmind-dir'
 * Use case: Change the path of Maxmind database files to writable directory
 *
 * @param  string $dir original directory of database files
 * @return string $dir replaced directory of database files
 */
function my_maxmind_dir( $dir ) {
	$upload = wp_upload_dir();
	return $upload['basedir'];
}

add_filter( 'ip-location-block-maxmind-dir', 'my_maxmind_dir' );


/**
 * Example 14: Usage of 'ip-location-block-maxmind-zip-ipv[46]'
 * Use case: Replace Maxmind database files to city edition
 *
 * @param  string $url original url to zip file
 * @return string $url replaced url to zip file
 */
function my_maxmind_ipv4( $url ) {
	return 'https://geolite.maxmind.com/download/geoip/database/GeoLiteCity.dat.gz';
}

function my_maxmind_ipv6( $url ) {
	return 'https://geolite.maxmind.com/download/geoip/database/GeoLiteCityv6-beta/GeoLiteCityv6.dat.gz';
}

add_filter( 'ip-location-block-maxmind-zip-ipv4', 'my_maxmind_ipv4' );
add_filter( 'ip-location-block-maxmind-zip-ipv6', 'my_maxmind_ipv6' );


/**
 * Example 15: Usage of 'ip-location-block-ip2location-path'
 * Use case: Change the path to IP2Location database files
 *
 * @param  string $path original path to database files
 * @return string $path replaced path to database files
 */
function my_ip2location_path( $path ) {
	return WP_PLUGIN_DIR . '/ip2location-tags/IP2LOCATION-LITE-DB1.IPV6.BIN';
}

add_filter( 'ip-location-block-ip2location-path', 'my_ip2location_path' );


/**
 * Example 16: Backup validation logs to text files
 * Use case: Keep verification logs selectively to text files
 *
 * @param  string $hook 'comment', 'xmlrpc', 'login', 'admin'
 * @param  string $dir default path where text files should be saved
 * @return string should be absolute path out of the public_html.
 */
function my_backup_dir( $dir, $hook ) {
	if ( in_array( $hook, array( 'login', 'admin' ) ) )
		return '/absolute/path/to/';
	else
		return null;
}

add_filter( 'ip-location-block-backup-dir', 'my_backup_dir', 10, 2 );


/**
 * Example 17: Usage of 'IP_Location_Block::get_geolocation()'
 * Use case: Get geolocation of visitor's ip address with latitude and longitude
 *
 */
function my_geolocation() {
	/**
	 * IP_Location_Block::get_geolocation(
	 *    $ip = NULL, $providers = array(), $callback = 'get_country'
	 * );
	 *
	 * @param string $ip IP address / default: $_SERVER['REMOTE_ADDR']
	 * @param array  $providers list of providers / ex: array( 'ipinfo.io' )
	 * @param string $callback geolocation function / ex: 'get_location'
	 * @return array country code and so on
	 */
	$geolocation = IP_Location_Block::get_geolocation();

	/**
	 * 'ip'       => string   validated ip address
	 * 'auth'     => int      authenticated or not
	 * 'code'     => string   country code
	 * 'time'     => unsinged int processing time
	 * 'provider' => string   IP geolocation service provider
	 */
	var_dump( $geolocation );

	if ( isset( $geolocation['errorMessage'] ) ) {
		echo 'error on getting geolocation'; // error handling
	}
}


/**
 * Example 18: Usage of 'ip-location-block-record-logs'
 * Use case: Prevent recording logs when it requested from own country
 *
 * @param  int    $record   0:none 1:blocked 2:passed 3:unauth 4:auth 5:all
 * @param  string $hook     'comment', 'xmlrpc', 'login' or 'admin'
 * @param  array  $validate the result of validation which contains:
 *  'ip'       => string    ip address
 *  'auth'     => int       authenticated (>= 1) or not (0)
 *  'code'     => string    country code
 *  'time'     => unsinged  processing time for examining the country code
 *  'provider' => string    IP geolocation service provider
 *  'result'   => string    'passed' or the reason of blocking
 * @return int    $record   modified condition
 */
function my_record_logs( $record, $hook, $validate ) {
	/* if request is from my country and passed, then no record */
	if ( 'JP' === $validate['code'] && 'passed' === $validate['result'] )
		$record = 0;

	return $record;
}

add_filter( 'ip-location-block-record-logs', 'my_record_logs', 10, 3 );

/**
 * Example 19: Specify the domain name server for reverse DNS lookup
 * Use case: Speed up reverse DNS lookup when on-site service is slow
 *
 * @param  array $servers empty array
 * @return array $servers array of primary and secondary servers
 */
function my_gethostbyaddr( $servers ) {
	return array( '1.1.1.1', '1.0.0.1' ); // APNIC  public DNS (faster)
//	return array( '8.8.8.8', '8.8.4.4' ); // Google public DNS (slower)
}

add_filter( 'ip-location-block-dns', 'my_gethostbyaddr' );
